[REM3-426] Add system property to configure default values for MAX_IN…

bmaxwell opened the pull request:
…BOUND_CHANNELS and MAX_OUTBOUND_CHANNELS

https://issues.redhat.com/browse/REM3-426

jbaesner committed:
[REM3-426] Add system property to configure default values for MAX_INBOUND_CHANNELS and MAX_OUTBOUND_CHANNELS

dmlloyd commented:
Given that we should never see more than 3 or 4 channels, what are the circumstances around needing to adjust this? It is not a capacity tunable.

bmaxwell commented:
@dmlloyd , @rachmato could explain better, but there is some issue where each deployment requires a channel and a large number of deployments runs out of channels.

rachmatowicz commented:
Hi @dmlloyd 

When we deploy a deployment containing an EJB client, a Contextual<EJBClientContext> is created  to act as a per-deployment EJBClientContext. That EJBCLientContext creates a new RemoteEJBReceiver, which in turn creates a new ClientServiceHandler which imposes a maximum of one channel per connection for all EJB clients using that RemoteEJBReceiver.

Each deployment to the same target server shares the same connection (via the connection sharing mechanism) , but each has their own ClientServiceHandler. Two deployments, two ClientServiceHandler instances, resulting in two channels of the same type (e.g. jboss.ejb) on the connection. 

We (with Rio) looked at sharing the ClientServiceHandler across deployments, but this broke encapsualtion of the Contextuals, IIRC.

dmlloyd commented:
OK, that's fair. Maybe we should bump up the default to, say, 400 - but add a comment explaining why, because my memory for this kind of thing is ~3 months at best. :-)

jbaesner commented on src/main/java/org/jboss/remoting3/RemotingOptions.java:
```diff
@@ -40,8 +42,16 @@
+    private static final String MAX_OUTBOUND_CHANNELS_KEY = "remoting.endpoint.max.outbound.channels";
+
+    private static final int HARD_CODED_DEFAULT_MAX_INBOUND_CHANNELS = 40;
+    private static final int HARD_CODED_DEFAULT_MAX_OUTBOUND_CHANNELS = 40;
```
Having this configurable enables users to adjust the values if the defaults do not meet their needs, but given we already have users exceeding the default limit and @dmlloyd suggestion to increase the defaults, shouldn't we change the default (while keeping it configurable) with this changed directly?

bmaxwell commented:
@jbaesner  @dmlloyd @rachmatowicz , I updated the PR so it defaults to 400 as per the comment.  And then if someone happens to overrun that, then they could set the system property.  I added a comment in the code referencing REM3-426 and added the notes above on the REM3-426 jira for future reference.

jbaesner commented:
> @jbaesner @dmlloyd @rachmatowicz , I updated the PR so it defaults to 400 as per the comment. And then if someone happens to overrun that, then they could set the system property. I added a comment in the code referencing REM3-426 and added the notes above on the REM3-426 jira for future reference.

LGTM
